ZEP-XXX-journal-record-format.md

```diff
@@ -0,0 +1,239 @@
+---
+title: Journal record format
+authors:
+  - Deepthi Akkoorath
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: yyyy-mm-dd
+last-updated: yyyy-mm-dd
+status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced
+
+---
+
+# Summary
+[summary]: #summary
+
+
+This document describes the format of a journal record as defined in the upcoming release Zeebe 1.0.0.
+
+In this document we describe fields added by the raft and the journal to the record.
+The fields added by the process engine is out of scope for this document.
+
+Different fields of a journal record is encoded using [SBE](https://github.com/real-logic/simple-binary-encoding). Please see Zeebe repository to find the SBE schema.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+
+## Journal layout
+The journal contains one or more segments. Each segment is a file that contains a sequence of records ordered by index.
+
+
+TODO: JournalSegmentDescriptor - first 64 bytes of a segment
+
+
+## Journal record format
+
+Notations: X<T> indicates the field X is of type T. For example checksum<int64> is a field named checksum of type int64 (long).
+X<SBE(Y)> indicates field X is an SBE (Simple Binary Encoding) message defined by the schema Y.
+
+
+A journal record in a segment has the following format.
+
+```
++-----------------------------------------------------------
+| frame<byte> |  metadata<SBE(Metadata)> | data<SBE(Data)> |
++-----------------------------------------------------------
+```
+
+### Header
+A `frame` is of size 1 byte. The `frame` can take following values:  
+0 : End of File  (EOF)
+1 : Version 1
+
+If the value of the frame indicates EOF, then the following bytes do not contain a valid journal record.
+Otherwise a valid record must exist.
+Any non-zero value indicates the version of the record.
+
+A journal record consists of two parts. The first one is `metadata` and the second one is `data`.
+Note that the record itself is not an SBE message.
+
+### Metadata
+
+Metadata consists of metadata of the data. `checksum` represents the CRC32 checksum of the data. The `length` represents the number of bytes in data. A journal reader can use the checksum to verify if data is corrupted, before it is read.
+
+```
++ --------------------------------
+| checksum<int64> | length<int32> |
```

Comment:
unsigned int is interpreted as long in the generated code. Then we have to cast to int always because everywhere length is expected to be int. Besides, there is not much advantage making it unsigned. 2^31 bytes is already too big for a record.